.github/scripts/validateActionsAndWorkflows.sh

```diff
@@ -94,4 +85,123 @@ fi
+###############################################################################
+title 'Checking for mutable action references...'
+
+# Find yaml files in the `.github` directory
+yamlFiles="$(find . -type f \( -name "*.yml" -o -name "*.yaml" \))"
+
+# Parse a yaml file, looking for action usages
+# Disabling shellcheck because this function is invoked by name and is reachable
+# shellcheck disable=SC2317
+extractActionsFromYaml() {
+  # Search for "uses:" in the yaml file
+  local usesLines
+  usesLines="$(grep --no-filename 'uses:' "$1")"
+
+  # Ignore files without external action usages
+  if [[ -z "$usesLines" ]]; then
+    return 0
+  fi
+
+  # Normalize: remove leading -
+  usesLines="${usesLines//- uses:/uses:}"
+
+  # Normalize: remove quotes
+  usesLines="${usesLines//\"/ }"
+  usesLines="${usesLines//\'/ }"
+
+  # Normalize: remove carriage returns
+  usesLines="${usesLines//\\r/ }"
+
+  # Grab action names and refs from uses lines.
+  # At this point, these lines look like "uses: myAction@ref", so `awk '{print $2}'` just grabs the second word from each line.
+  local actionsWithRefs
+  actionsWithRefs="$(echo "$usesLines" | awk '{print $2}')"
+  echo "$actionsWithRefs"
+  echo $'\n'
+}
+
+# Parse all yaml files in parallel to find all action usages
+while IFS= read -r yamlFile; do
+  run_async extractActionsFromYaml "$yamlFile"
+done <<< "$yamlFiles"
+actionUsages="$(await_async_commands)"
+
+# De-dupe and sort action usages
+actionUsages="$(echo "$actionUsages" | grep -vE '^$' | sort | uniq)"
+
+info 'All action usages...'
+echo "$actionUsages"
+echo
+
+# Ignore any local action usages, callable workflows, or Expensify-owned actions
+actionUsages="$(echo "$actionUsages" | grep -vE "^(.github|Expensify/)")"
+
+info 'Untrusted action usages...'
+echo "$actionUsages"
+echo
+
+# Next, we'll check all the actions we found to make sure they're immutable.
+# We're using a temp file instead of a variable so we can write to it from a subprocess (i.e: a command run in the background)
+mutableActionUsages="$(mktemp)"
+
+# Given an action name with a ref (actions/checkout@v2), check the actual repo to make sure it's an immutable commit hash
+# and not secretly a tag or branch that looks like a commit hash.
+# If it's mutable, collect it into the mutableActionUsages file.
+# shellcheck disable=SC2317
+verifyActionRefIsImmutable() {
+  local actionWithRef="$1"
+
+  # Everything before the @
+  local action
+  action="${1%@*}"
+
+  # Everything after the @
+  local ref
+  ref="${1##*@}"
+
+  # Check if the ref looks like a commit hash (40-character hexadecimal string)
+  if [[ ! "$ref" =~ ^[0-9a-f]{40}$ ]]; then
+    # Ref does not look like a commit hash, and therefore is probably mutable
+    echo "$actionWithRef" >> "$mutableActionUsages"
+    return 1
+  fi
+
+  # Ref looks like a commit hash, but we need to check the remote to make sure it's not a tag or a branch
+  local repoURL
+  repoURL="git@github.com:${action}.git"
```

Comment:
Found an action for which this doesn't work: `gradle/actions/setup-gradle`. So need to mark this as WIP

Reply:
we fixed this